src/analysis/vector/qgsgeometrysnapper.cpp

```diff
@@ -402,14 +350,34 @@ QgsSnapIndex::SnapItem *QgsSnapIndex::getSnapItem( const QgsPoint &pos, double t
   int colEnd = std::floor( ( pos.x() + tol - mOrigin.x() ) / mCellSize );
   int rowEnd = std::floor( ( pos.y() + tol - mOrigin.y() ) / mCellSize );
 
-  rowStart = std::max( rowStart, mRowsStartIdx );
-  rowEnd = std::min( rowEnd, mRowsStartIdx + mGridRows.size() - 1 );
+  GEOSContextHandle_t geosctxt = QgsGeos::getGEOSHandler();
+
+  GEOSCoordSequence *coord = GEOSCoordSeq_create_r( geosctxt, 5, 2 );
+#if GEOS_VERSION_MAJOR>3 || GEOS_VERSION_MINOR>=8
+  GEOSCoordSeq_setXY_r( geosctxt, coord, 0, rowStart, colStart );
+  GEOSCoordSeq_setXY_r( geosctxt, coord, 1, rowEnd, colStart );
+  GEOSCoordSeq_setXY_r( geosctxt, coord, 2, rowEnd, colEnd );
+  GEOSCoordSeq_setXY_r( geosctxt, coord, 3, rowStart, colEnd );
+  GEOSCoordSeq_setXY_r( geosctxt, coord, 4, rowStart, colStart );
+#else
+  GEOSCoordSeq_setX_r( geosctxt, coord, 0, rowStart );
+  GEOSCoordSeq_setY_r( geosctxt, coord, 0, colStart );
+  GEOSCoordSeq_setX_r( geosctxt, coord, 1, rowEnd );
+  GEOSCoordSeq_setY_r( geosctxt, coord, 1, colStart );
+  GEOSCoordSeq_setX_r( geosctxt, coord, 2, rowEnd );
+  GEOSCoordSeq_setY_r( geosctxt, coord, 2, colEnd );
+  GEOSCoordSeq_setX_r( geosctxt, coord, 3, rowStart );
+  GEOSCoordSeq_setY_r( geosctxt, coord, 3, colEnd );
+  GEOSCoordSeq_setX_r( geosctxt, coord, 4, rowStart );
+  GEOSCoordSeq_setY_r( geosctxt, coord, 4, colStart );
+#endif
+
+  geos::unique_ptr searchArea( GEOSGeom_createPolygon_r( geosctxt, GEOSGeom_createLinearRing_r( geosctxt, coord ), nullptr, 0 ) );
 
   QList<SnapItem *> items;
-  for ( int row = rowStart; row <= rowEnd; ++row )
-  {
-    items.append( mGridRows[row - mRowsStartIdx].getSnapItems( colStart, colEnd ) );
-  }
+  struct _GEOSQueryCallbackData callbackData;
+  callbackData.list = &items;
+  GEOSSTRtree_query_r( geosctxt, mSTRTree, searchArea.get(), _GEOSQueryCallback, &callbackData );
```

Comment:
You don't actually need to construct a polygon for the search area, because GEOS is just going to take the envelope. Constructing a two-point linestring would be a little more efficient.

Reply:
@dbaston , very good point, I shall adjust. 

I was actually going to reach out to you in the coming hours about this neat use of GEOS STR Tree. There's one showstopper ATM: this solution works flawlessly on GEOS 3.9 but fails completely when used against GEOS 3.8.0 (as per the failing geometry snapper test in our CI running Ubuntu 20.04).

Is there something I'm doing here that GEOS 3.8 doesn't like?

And while we're at it, since I've not had the good fortune to interact with you before, thanks for your work on GEOS :)

Reply:
Hmm, I don't see anything that would be a problem with 3.8 (or any other version, for that matter.) Does the test failure involve a crash or an unexpected result?

Reply:
@dbaston , no crash, but the querying of the index doesn't appear to return any matches (on line https://github.com/qgis/QGIS/pull/42781/files#diff-85f582115e6e43bb3410dea47308187c69ddece5c0621adf3f3aef770ef3cb60R324 )

I was hoping 3.8.1 would work, but I tried on my secondary machine this morning with no luck, the test that works just fine on GEOS 3.9 fails on GEOS 3.8.1.

Reply:
We did switch to a new STRtreein 3.9, but the previous one was in place for a long time and I'm not aware of any issues with it. Is the failing test in Is the failing test in https://github.com/qgis/QGIS/blob/2d1aa68f0d044f2aced7ebeca8d2fa6b754ac970/tests/src/analysis/testqgsgeometrysnapper.cpp ? Which case is failing?

Reply:
@dbaston, did the previous tree code not like points as envelop?

Reply:
Not intentionally, but it's an interesting idea. Try buffering by some tiny amount and seeing if that fixes it?

Reply:
@dbaston , no luck, I tried using a diagonal line to increase envelop for the query geometry, as well as the individual items added to the tree. 

We might have to #ifdef this huge optimization for QGIS built against GEOS 3.9. That'd leave a lot of users behind for the time being.

Reply:
Which test case is failing?

Reply:
@dbaston , all of these : https://github.com/qgis/QGIS/blob/master/tests/src/analysis/testqgsgeometrysnapper.cpp